pkg/core/klinedriver/tick_klinedriver_test.go

```diff
@@ -0,0 +1,160 @@
+package klinedriver
+
+import (
+	"testing"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/stretchr/testify/assert"
+)
+
+type KLineCollector struct {
+	KLines []*types.KLine
+}
+
+func (c *KLineCollector) EmitKLine(kline types.KLine) {
+	c.KLines = append(c.KLines, &kline)
+}
+func (c *KLineCollector) EmitKLineClosed(kline types.KLine) {
+	c.KLines = append(c.KLines, &kline)
+}
+
+// TestTickKLineDriver tests
+// It also demonstrates how to use the driver in backtesting
+func TestTickKLineDriver(t *testing.T) {
+	symbol := "NOTEXIST"
+	t.Run("TestTrades", func(t *testing.T) {
+		tickStartTime, _ := time.Parse(time.RFC3339, "2023-10-01T01:35:00Z")
+		driver := NewTickKLineDriver(symbol, time.Second*10)
+		driver.Subscribe("1m")
+		kLineCollector := &KLineCollector{}
+		driver.SetKLineEmitter(kLineCollector)
+		driver.SetRunning(tickStartTime)
+		for _, trade := range []types.Trade{
+			{
+				Symbol: symbol,
+				Price:  fixedpoint.NewFromFloat(100),
+				Time:   types.Time(tickStartTime.Add(5 * time.Second)),
+			},
+			{
+				Symbol: symbol,
+				Price:  fixedpoint.NewFromFloat(200),
+				Time:   types.Time(tickStartTime.Add(15 * time.Second)),
+			},
+			{
+				Symbol: symbol,
+				Price:  fixedpoint.NewFromFloat(20),
+				Time:   types.Time(tickStartTime.Add(30 * time.Second)),
+			},
+		} {
+			driver.AddTrade(trade)
+		}
+		driver.ProcessTick(tickStartTime.Add(30 * time.Second).Add(123 * time.Millisecond))
+		assert.Equal(t, 1, len(kLineCollector.KLines))
+		kline := kLineCollector.KLines[0]
+		assert.Equal(t, fixedpoint.NewFromFloat(100), kline.Open)
+		assert.Equal(t, fixedpoint.NewFromFloat(20), kline.Close)
+		assert.Equal(t, fixedpoint.NewFromFloat(200), kline.High)
+		assert.Equal(t, fixedpoint.NewFromFloat(20), kline.Low)
+		assert.Equal(t, uint64(3), kline.NumberOfTrades)
+		kLineCollector.KLines = nil
+		driver.AddTrade(types.Trade{
+			Symbol: symbol,
+			Price:  fixedpoint.NewFromFloat(50),
+			Time:   types.Time(tickStartTime.Add(55 * time.Second)),
+		})
+		driver.ProcessTick(tickStartTime.Add(time.Minute).Add(123 * time.Millisecond))
+		// 1 closed kline, 1 open kline
+		assert.Equal(t, 2, len(kLineCollector.KLines))
+		// closed kline
+		kline = kLineCollector.KLines[0]
+		assert.True(t, kline.Closed)
+		assert.Equal(t, fixedpoint.NewFromFloat(100), kline.Open)
+		assert.Equal(t, fixedpoint.NewFromFloat(50), kline.Close)
+		assert.Equal(t, fixedpoint.NewFromFloat(200), kline.High)
+		assert.Equal(t, fixedpoint.NewFromFloat(20), kline.Low)
+		assert.Equal(t, uint64(4), kline.NumberOfTrades)
+		// open kline
+		kline = kLineCollector.KLines[1]
+		assert.False(t, kline.Closed)
+		assert.Equal(t, fixedpoint.NewFromFloat(50), kline.Open)
+		assert.Equal(t, fixedpoint.NewFromFloat(50), kline.Close)
+		assert.Equal(t, fixedpoint.NewFromFloat(50), kline.High)
+		assert.Equal(t, fixedpoint.NewFromFloat(50), kline.Low)
+		assert.Equal(t, uint64(0), kline.NumberOfTrades)
```

Comment:
ditto

Reply:
Opps, I forgot to add the checks on start/end time.
Will update the test.